internal/provider/services/storage/storage.go

```diff
@@ -0,0 +1,235 @@
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/terrariumcloud/terrarium/internal/storage"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	DefaultBucketName                          = "terrarium-providers"
+	DefaultStorageServiceDefaultEndpoint       = "storage:3001"
+	DefaultChunkSize                     int64 = 64 * 1024 // 64 KB
+)
+
+var (
+	BucketName             = DefaultBucketName
+	StorageServiceEndpoint = DefaultStorageServiceDefaultEndpoint
+	ChunkSize              = DefaultChunkSize
+
+	BucketInitializationError = status.Error(codes.Unknown, "Failed to initialize bucket for storage.")
+	DownloadSourceZipError    = status.Error(codes.Unknown, "Failed to download source zip.")
+	SendSourceZipError        = status.Error(codes.Unknown, "Failed to send source zip.")
+	SendShasumError           = status.Error(codes.Unknown, "Failed to send shasum file.")
+	ContentLengthError        = status.Error(codes.Unknown, "Failed to read correct content length.")
+	DownloadShasumError       = status.Error(codes.Unknown, "Failed to download shasum.")
+)
+
+type StorageService struct {
+	services.UnimplementedStorageServer
+	Client     storage.AWSS3BucketClient
+	BucketName string
+	Region     string
+}
+
+// Registers StorageService with grpc server
+func (s *StorageService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) error {
+	if err := storage.InitializeS3Bucket(s.BucketName, s.Region, s.Client); err != nil {
+		log.Println(err)
+		return BucketInitializationError
+	}
+
+	services.RegisterStorageServer(grpcServer, s)
+
+	return nil
+}
+
+// Download Source Zip from storage
+func (s *StorageService) DownloadProviderSourceZip(request *services.DownloadSourceZipRequest, server services.Storage_DownloadProviderSourceZipServer) error {
+
+	log.Println("Downloading source zip.")
+
+	ctx := server.Context()
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("provider.name", request.GetProvider().GetName()),
+		attribute.String("provider.version", request.GetProvider().GetVersion()),
+		attribute.String("provider.os", request.GetProvider().GetOs()),
+		attribute.String("provider.arch", request.GetProvider().GetArch()),
+	)
+
+	providerAddress := strings.Split(request.Provider.GetName(), "/")
+	filename := fmt.Sprintf("terraform-provider-%s_%s_%s_%s.zip", providerAddress[1], request.GetProvider().GetVersion(), request.GetProvider().GetOs(), request.GetProvider().GetArch())
+	fileLocation := services.ResolveS3Locations(request.Provider.GetName(), request.Provider.GetVersion(), filename)
+
+	in := &s3.GetObjectInput{
+		Bucket: aws.String(BucketName),
+		Key:    aws.String(fileLocation),
+	}
+
+	out, err := s.Client.GetObject(ctx, in)
+	if err != nil {
+		span.RecordError(err)
+		log.Println(err)
```

Comment:
These log statements aren't very useful as they don't explain from what operation the error occurred.
Consider either enhancing the message or removing

Reply:
Thanks for the suggestion. I will consider enhancing them.